Add the V (vector) extension to the riscv64-linux-android target spec

chriswailes opened the pull request:
This feature has been enabled and tested internally in the Android project.

chriswailes committed:
Add the V (vector) extension to the riscv64-linux-android target spec

This feature has been enabled and tested internally in the Android
project.

workingjubilee commented:
This is a tier 3 target, so arbitrary fiddling with its spec is permitted, and it does not encode extensions like `gc` into its name, so there is no question of whether it should be `riscv64gcv-linux-android`. And, yk, it's the target maintainer.

But I do think we should do one thing.

@chriswailes Can you mention the ISA requirements in the platform support file? You can use "currently... at least..." kind of cagey terminology if you want to avoid committing to stuff, but if someone out there tries this on a random RV64 board, even if they had previously gotten Android working on it, it's likely to fail now, and some boards claim they support the V extension but actually support a pre-ratification version only which is not binary-compatible with the finalized version. It'd be nice to not ask people to gaze into the rustc source to find out why.

chriswailes committed:
Add documentation for the riscv64-android-linux target

This commit adds additional documentation describing the
features/extensions required by the riscv64-linux-android target.

chriswailes commented:
Sure thing!  Thanks for the great suggestion.  I debated being a little tricky with the list and doing something like:

```
* a(tomics)
* d(ouble-precision floating-point)
```

but eventually decided against it.  Let me know if you'd like any changes or additions to the documentation.

workingjubilee commented:
Thank you!

I know it's common for SBC makers and the like to carry a bunch of patches for a while until they get stuff upstreamed, so if anyone is already in the process of putting Rust and Android on RV64, hopefully this makes life easier for them.

@bors r+

bors commented:
:pushpin: Commit 899f81fcb8a4ead2524d5728cb40518b085496bf has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 899f81fcb8a4ead2524d5728cb40518b085496bf -->
<!-- homu: {"type":"Approved","sha":"899f81fcb8a4ead2524d5728cb40518b085496bf","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 899f81fcb8a4ead2524d5728cb40518b085496bf with merge 04287982ef974631b34b2cf0ee19fc220d89e707...
<!-- homu: {"type":"BuildStarted","head_sha":"899f81fcb8a4ead2524d5728cb40518b085496bf","merge_sha":"04287982ef974631b34b2cf0ee19fc220d89e707"} -->

taiki-e commented on compiler/rustc_target/src/spec/riscv64_linux_android.rs:
```diff
@@ -9,7 +9,7 @@ pub fn target() -> Target {
             code_model: Some(CodeModel::Medium),
             cpu: "generic-rv64".into(),
-            features: "+m,+a,+f,+d,+c,+Zba,+Zbb,+Zbs".into(),
+            features: "+m,+a,+f,+d,+c,+Zba,+Zbb,+Zbs,+V".into(),
```
I think these need to be lower case.

```suggestion
            features: "+m,+a,+f,+d,+c,+zba,+zbb,+zbs,+v".into(),
```

https://github.com/rust-lang/rust/blob/130ff8cb6c3d62ed66daf652cbb5323d3f93c4fc/compiler/rustc_codegen_ssa/src/target_features.rs#L255-L257

The current one generates "recognized feature" warnings.

```
'+Zba' is not a recognized feature for this target (ignoring feature)
'+Zbb' is not a recognized feature for this target (ignoring feature)
'+Zbs' is not a recognized feature for this target (ignoring feature)
'+V' is not a recognized feature for this target (ignoring feature)
```

https://godbolt.org/z/cEq1d6vK6

workingjubilee replied:
...Wow that's annoying.

chriswailes replied:
Done

workingjubilee replied:
Thank you both.

workingjubilee commented:
Eh, really?
@bors r-

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6503779649/job/17666394527)
Build commit: 04287982ef974631b34b2cf0ee19fc220d89e707 (`04287982ef974631b34b2cf0ee19fc220d89e707`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6503779649/job/17666394527"},"merge_sha":"04287982ef974631b34b2cf0ee19fc220d89e707"} -->

chriswailes committed:
Lowercase the feature flags for riscv64-linux-android

workingjubilee commented:
rustc... really should have caught that on its own.

Anyway,
@bors r+

bors commented:
:pushpin: Commit 166c3534849c88e38ca730457970d916a9d31f1a has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 166c3534849c88e38ca730457970d916a9d31f1a -->
<!-- homu: {"type":"Approved","sha":"166c3534849c88e38ca730457970d916a9d31f1a","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
